### PR TITLE
Simplify signer set

### PR DESCRIPTION
**Pull request by pfletcherhill**
Implements idea 2 from https://www.notion.so/farcasterxyz/Signer-set-cut-back-ideas-afb8ca2b1d7543a6ad1164425b264c14

Summary of changes:
- Add data structure (`CustodyAddEvent`) representing FIR events (i.e. transfer and register)
- Add `CustodyRemoveAll` message type, which removes all custody addresses prior to the custody address signing this message
- Only allow custody addresses to add/remove delegate signers
- Drop vertices/edges data structure in favor of LWW set for delegate signers
- Add comment at top of SignerSet with explanation of sets and conflict resolution logic (custody block number, remove > add, custody address lexicographical order, message hash lexicographical order)
- Add [tiny-typed-emitter](https://github.com/binier/tiny-typed-emitter) library for typing event emitters
- Make SignerSet an event emitter and specify supported events in `SignerSetEvents`
- When a custody address is removed, drop all messages signed by that address and emit resulting events (i.e. `removeSigner` when SignerAdd message is dropped)
- Re-write SignerSet tests and check full set state (`custodyAdds`, `custodyRemoves`, `signerAdds`, `signerRemoves`, and ordered list of emitted events)

### CODE REVIEW

File: src/sets/signerSet.ts [hunk 1]
```diff
@@ -11,79 +12,75 @@ import { hashCompare } from '~/utils';
...
+export type SignerSetEvents = {
+  addCustody: (custodyAddress: string) => void;
+  removeCustody: (custodyAddress: string) => void;
+  addSigner: (signerKey: string) => void;
+  removeSigner: (signerKey: string) => void;
+};
+
+class SignerSet extends TypedEmitter<SignerSetEvents> {
+  private _custodyAdds: Map<string, CustodyAddEvent>;
+  private _custodyRemoves: Map<string, CustodyRemoveAll>;
```
Comment (varunsrin) [inline src/sets/signerSet.ts:40]: nit: i've been using `adds/rems` in some places and here we use `adds/removes`. slight bias towards `adds/rems` since its more concise and overlaps with crdt language.

File: src/sets/signerSet.ts [hunk 2]
```diff
@@ -11,79 +12,75 @@ import { hashCompare } from '~/utils';
...
 
-  getDelegates(): string[] {
-    return Array.from(this._vertexAdds);
+  getAllSigners(): Set<string> {
+    return new Set([...this.getCustodyAddresses(), ...this.getDelegateSigners()]);
   }
 
-  getAllDelegates(): string[] {
-    return Array.from(this._vertices);
+  lookup(signer: string) {
```
Comment (varunsrin) [inline src/sets/signerSet.ts:64]: nit: should we call these methods `isValid` , `isValidCustody` and `isValidDelegate` instead?

lookup describes the operation we are performing but not the context around it (looking up a currently valid signer)

File: src/sets/signerSet.ts [hunk 3]
```diff
@@ -11,79 +12,75 @@ import { hashCompare } from '~/utils';
...
-  constructEdgeKey(parentKey: string, childKey: string): string {
-    return [this.sanitizeKey(parentKey), this.sanitizeKey(childKey)].toString();
+  lookupSigner(signer: string) {
+    return this._signerAdds.get(signer);
   }
 
-  deconstructEdgeKey(edgeKey: string): { parentKey: string; childKey: string } {
-    const [parentKey, childKey] = edgeKey.split(',');
-    return { parentKey, childKey };
+  sanitizeKey(key: string): string {
```
Comment (varunsrin) [inline src/sets/signerSet.ts:76]: nit: what is the purpose of this alias? it is similar in length to the function it aliases and i'd argue that it's less clear about what it's doing.  initial reaction is that just calling `toLowerCase` everywhere makes understanding the code easier.

File: src/sets/signerSet.ts [hunk 4]
```diff
@@ -99,381 +96,224 @@ class SignerSet {
...
-  /**
-   * removeCustody moves a custody address from the custodyAdds set to the custodyRemoves set.
-   *
-   * @param custodyAddress - custody address to move to custodyRemoves set
+   * @param event - event from Farcaster ID Registry
    */
-  removeCustody(custodyAddress: string): Result<void, string> {
-    const sanitizedAddress = this.sanitizeKey(custodyAddress);
+  addCustody(event: CustodyAddEvent): Result<void, string> {
+    const sanitizedAddress = this.sanitizeKey(event.custodyAddress);
```
Comment (varunsrin) [inline src/sets/signerSet.ts:102]: nit: personally would prefer `lowerCaseAddr` to `sanitizedAddr` since it is clearer what we are doing with the address

File: src/sets/signerSet.ts [hunk 5]
```diff
@@ -11,79 +12,75 @@ import { hashCompare } from '~/utils';
...
-  deconstructEdgeKey(edgeKey: string): { parentKey: string; childKey: string } {
-    const [parentKey, childKey] = edgeKey.split(',');
-    return { parentKey, childKey };
+  sanitizeKey(key: string): string {
+    return key.toLowerCase();
   }
 
-  merge(message: SignerMessage): Result<void, string> {
-    if (this._messages.get(message.hash)) return ok(undefined); // Duplicate message
+  merge(message: SignerMessage | CustodyRemoveAll): Result<void, string> {
```
Comment (varunsrin) [inline src/sets/signerSet.ts:80]: it was slightly confusing that merge and addCustody operate on the same structures but have very different names. 

should merge also accept the addCustody events and route them to the right method? Then there's a single api to call into with any type of update message or event

File: src/sets/signerSet.ts [hunk 6]
```diff
@@ -99,381 +96,224 @@ class SignerSet {
...
+    // If custody exists in custodyRemoves
+    const existingCustodyRemove = this._custodyRemoves.get(sanitizedAddress);
+    if (existingCustodyRemove) {
+      const existingCustodyAddEvent = this._custodyAdds.get(this.sanitizeKey(existingCustodyRemove.signer));
 
-    // Fail if address does not exist in custodyAdds
-    if (!this._custodyAdds.has(sanitizedAddress)) return err('SignerSet.removeCustody: custodyAddress does not exist');
+      if (existingCustodyAddEvent) {
+        // If existing block number wins (same or greater block), no-op
+        if (existingCustodyAddEvent.blockNumber > event.blockNumber) return ok(undefined);
```
Comment (varunsrin) [inline src/sets/signerSet.ts:118]: did you mean `>=` here?

File: src/sets/signerSet.ts [hunk 7]
```diff
@@ -99,381 +96,224 @@ class SignerSet {
...
-
-  /**
-   * removeCustody moves a custody address from the custodyAdds set to the custodyRemoves set.
-   *
-   * @param custodyAddress - custody address to move to custodyRemoves set
+   * @param event - event from Farcaster ID Registry
    */
-  removeCustody(custodyAddress: string): Result<void, string> {
-    const sanitizedAddress = this.sanitizeKey(custodyAddress);
+  addCustody(event: CustodyAddEvent): Result<void, string> {
```
Comment (varunsrin) [inline src/sets/signerSet.ts:101]: writing out the logic as documentation will make this easier to review and update in the future. perhaps something like: 

1. if this address has never been seen before, add it 
2. if this address has been seen and exists in the add set, only add it if it is more recent than the last known one
3. if this address has been seen and exists in the rem set, then find the signer that authorized the removal, and if it is older than the block this address was added in, move it from the rem set to the add set

File: src/sets/signerSet.ts (hunk 1 above)
Comment (varunsrin) [reply src/sets/signerSet.ts:40]: nit: it might be worth adding a comment here that this maps *every* removed address to a *CustodyRemoveAll* message and that there may be duplicate  remove all messages

File: src/sets/signerSet.ts [hunk 8]
```diff
@@ -99,381 +96,224 @@ class SignerSet {
...
+          if (this.sanitizeKey(signerAdd.signer) === address) {
+            this._signerAdds.delete(signerKey);
+            this.emit('removeSigner', signerKey);
+          }
+        }
 
-    this._messages.set(message.hash, message);
-    return ok(undefined);
-  }
+        // Look through signerRemoves
```
Comment (varunsrin) [inline src/sets/signerSet.ts:163]: nit: would be helpful to elaborate on why for someone new to this

e..g `Remove all SignerRemoves that were authorized by a CustodyAddress that is now removed by this CustodyRemAll message`

File: src/sets/signerSet.ts [hunk 9]
```diff
@@ -99,381 +96,224 @@ class SignerSet {
...
+    const existingSignerAdd = this._signerAdds.get(signerKey);
+    if (existingSignerAdd) {
+      const existingCustodyAddEvent = this._custodyAdds.get(this.sanitizeKey(existingSignerAdd.signer));
+
+      if (existingCustodyAddEvent) {
+        // If existing block number wins, no-op
+        if (existingCustodyAddEvent.blockNumber > custodyAddEvent.blockNumber) return ok(undefined);
+
+        // If block numbers are the same
+        if (existingCustodyAddEvent.blockNumber === custodyAddEvent.blockNumber) {
```
Comment (varunsrin) [inline src/sets/signerSet.ts:212]: is it possible to create a contract that could shuffle around an id between two addresses multiple times in the same transaction? and would that cause a problem here?

Comment (varunsrin) [reply src/sets/signerSet.ts:212]: this logic is particularly hard to parse as a reader -- i think its just very tricky to understand and might need lots of docs

File: src/sets/signerSet.ts (hunk 3 above)
Comment (pfletcherhill) [reply src/sets/signerSet.ts:76]: Not a strong opinion either way, but I did it this way to future-proof it. If we want to change from lowercase to something else, we only have to change it once rather than looking at every `toLowerCase` call in the codebase to decide if it's relevant or not to sanitizing signer keys.

File: src/sets/signerSet.ts (hunk 5 above)
Comment (pfletcherhill) [reply src/sets/signerSet.ts:80]: Yeah, I was unsure about this too. I think it depends whether we want (1) hubs to sync custody add events (i.e. contract events) with each other or (2) hubs to only add custody addresses by syncing with the on-chain FIR. What do you think between those?

If we're okay having custody add events synced between hubs, then let's toss it into `merge`.

File: src/sets/signerSet.ts (hunk 6 above)
Comment (pfletcherhill) [reply src/sets/signerSet.ts:118]: Ah yeah, oops!

File: src/sets/signerSet.ts (hunk 7 above)
Comment (pfletcherhill) [reply src/sets/signerSet.ts:101]: [The notion doc](https://www.notion.so/farcasterxyz/Signer-set-cut-back-ideas-afb8ca2b1d7543a6ad1164425b264c14) has pseudocode, but I can make it simpler and add in the code as a comment as well.

File: src/sets/signerSet.ts (hunk 8 above)
Comment (pfletcherhill) [reply src/sets/signerSet.ts:163]: Yeah, I'll add more explanation in comments

**Comment by pfletcherhill** [comment]
Merging and will address comments in the signers branch

File: src/sets/signerSet.ts (hunk 9 above)
Comment (pfletcherhill) [reply src/sets/signerSet.ts:212]: Adding comments and test to confirm that a CustodyRemoveAll message will only remove custody addresses added in a prior block. Custody addresses added in the same block (regardless of the order Ethereum decides for them), will be preserved until a custody address removes them from a later block.
